src/services/DatasetService.js

```diff
@@ -1,58 +1,134 @@
 import datasette from './datasette.js'
 import logger from '../utils/logger.js'
+import { types } from '../utils/logging.js'
+import JSON5 from 'json5' // need to use this as spec is written with single quotes. need to pass this onto data designers team to fix
 import performanceDbApi from './performanceDbApi.js'
 
-async function getDatasetStatsForResourceId (dataset, resourceId) {
-  const sql = `
-    SELECT 'numberOfRecords' AS metric, COUNT(*) AS value
-    FROM
-      (
-        SELECT
-          *
-        FROM
-          fact_resource fr
-        WHERE
-          fr.resource = '${resourceId}'
-        GROUP BY
-          entry_number
-      )
-    UNION ALL
-    SELECT 'numberOfFieldsSupplied' AS metric, COUNT(*) AS value
-    FROM
-    (
-      SELECT
-        *
-      FROM
-        fact_resource fr
-      WHERE
-        fr.resource = '${resourceId}'
-    )`
-
-  const { formattedData } = await datasette.runQuery(sql, dataset)
+async function getColumnSummary (dataset, lpa) {
+  const sql = `select * from column_field_summary
+    where resource != ''
+    and pipeline = '${dataset}'
+    AND organisation = '${lpa}'
+    limit 1000`
+
+  const { formattedData } = await datasette.runQuery(sql, 'performance')
 
   return formattedData
 }
 
-export async function getDatasetStats (dataset, lpa) {
-  try {
-    const stats = {}
-    const { resource: resourceId } = await performanceDbApi.getLatestResource(lpa, dataset)
-    const metrics = await getDatasetStatsForResourceId(dataset, resourceId)
+async function getFieldStats (lpa, dataset) {
+  const columnSummary = await datasetService.getColumnSummary(dataset, lpa)
+  const specifications = await datasetService.getSpecifications()
+  if (!(dataset in specifications)) {
+    logger.warn(`services/datasetService.getFieldStats(): cannot find specification for dataset: ${dataset}`, { type: types.app, dataset })
+    return null
+  }
+
+  const datasetSpecification = specifications[dataset]
+
+  const matchingFields = columnSummary[0].matching_field.split(',')
+  const nonMatchingFields = columnSummary[0].non_matching_field.split(',')
+  const allFields = [...matchingFields, ...nonMatchingFields]
+
+  const numberOfFieldsSupplied = datasetSpecification.fields.map(field => field.field).reduce((acc, current) => {
+    return allFields.includes(current) ? acc + 1 : acc
+  }, 0)
+
+  const numberOfFieldsMatched = datasetSpecification.fields.map(field => field.field).reduce((acc, current) => {
+    return matchingFields.includes(current) ? acc + 1 : acc
+  }, 0)
+
+  const numberOfExpectedFields = datasetSpecification.fields.length
+
+  return {
+    numberOfFieldsSupplied,
+    numberOfFieldsMatched,
+    numberOfExpectedFields
+  }
+}
+
+async function getSpecifications () {
```

Comment:
Again, `fetch[One|Many]` and move the data wrangling to a `prepareTemplateParamsFor*` which should be a pure function when possible (see `docs/Middleware Guidelines.md`). This way we can test it without mocking network calls. Obviously, It doesn't have be one giant fn,  add more specialised fns where necessary and call them from `prepareTemplateParams*`